ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WriteParticipantPruningService.scala

```diff
@@ -0,0 +1,25 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.v1
+
+import java.util.concurrent.CompletionStage
+
+/** An interface to prune participant ledger updates to manage participant ledger space and enable GDPR-style
+  * right-to-be-forgotten support. */
+trait WriteParticipantPruningService {
+
+  /** Prune the participant ledger specifying the offset up to which participant ledger events can be removed.
+    *
+    * As this interface applies only to the local participant unlike other administrator services, returns a
+    * (completion stage of a) PruningResult rather than a SubmissionResult.
+    *
+    * Ledgers that do not elect to support participant pruning, return NotPruned(Status.UNIMPLEMENTED). Returning an
+    * error also keeps the ledger api server from pruning its index.
```

Comment:
As noted above, we should probably talk more about failed pruning operations. Maybe also advise WriteService implementations to perform any destructive operations atomically (e.g., in one SQL transaction).

Reply:
I have written up a section on fault tolerance taking your points as inspiration. Atomic-SQL transactions are one model and another one is breaking up pruning into idempotent steps in such a way that they lend themselves to reaching eventual consistency upon failures.